File: lib/git.ml
let is_in_repo () =
  match
    Exec.read_cmd ~quiet_err:true "git"
      [ "rev-parse"; "--is-inside-work-tree" ]
  with
  | Ok _ -> true
  | Error _ -> false

let is_tree_clean () =
  match Exec.read_cmd ~quiet_err:true "git" [ "status"; "--porcelain" ] with
  | Ok s -> String.length s = 0
  | Error _ -> false

let resolve_ref ref = Exec.read_cmd "git" [ "rev-parse"; ref ]

let pipe_git_archive sha consumer_prog consumer_args =
  let pipe_r, pipe_w = Unix.pipe ~cloexec:true () in
  let archive_pid =
    Unix.create_process "git"
      [| "git"; "archive"; "--format=tar"; sha |]
      Unix.stdin pipe_w Unix.stderr
  in
  Unix.close pipe_w;
  let consumer_argv = Array.of_list (consumer_prog :: consumer_args) in
  let consumer_pid =
    Unix.create_process consumer_prog consumer_argv pipe_r Unix.stdout
      Unix.stderr
  in
  Unix.close pipe_r;
  let _, consumer_status = Unix.waitpid [] consumer_pid in
  let _, archive_status = Unix.waitpid [] archive_pid in
  match consumer_status, archive_status with
  | Unix.WEXITED 0, Unix.WEXITED 0 -> Ok ()
  | _ -> Error "extraction failed"

let extract_local sha dir =
  Exec.mkdir_p dir;
  match pipe_git_archive sha "tar" [ "-C"; dir; "-x" ] with
  | Error _ as e -> e
  | Ok () ->
    let chmod_pid =
      Unix.create_process "chmod"
        [| "chmod"; "-R"; "u+w"; dir |]
        Unix.stdin Unix.stdout Unix.stderr
    in
    let _, status = Unix.waitpid [] chmod_pid in
    (match status with
     | Unix.WEXITED 0 -> Ok ()
     | _ -> Error "chmod failed")

let extract_remote sha host dir =
  let remote_cmd =
    Printf.sprintf "mkdir -p '%s' && tar -C '%s' -x && chmod -R u+w '%s'" dir
      dir dir
  in
  let pipe_r, pipe_w = Unix.pipe ~cloexec:true () in
  let archive_pid =
    Unix.create_process "git"
      [| "git"; "archive"; "--format=tar"; sha |]
      Unix.stdin pipe_w Unix.stderr
  in
  Unix.close pipe_w;
  let ssh_pid =
    Unix.create_process "ssh"
      [| "ssh"; host; remote_cmd |]
      pipe_r Unix.stdout Unix.stderr
  in
  Unix.close pipe_r;
  let _, ssh_status = Unix.waitpid [] ssh_pid in
  let _, archive_status = Unix.waitpid [] archive_pid in
  match ssh_status, archive_status with
  | Unix.WEXITED 0, Unix.WEXITED 0 -> Ok ()
  | _ -> Error "remote extraction failed"

File: cmd/localci/git.go
// Git operations: repo detection, clean-tree check, HEAD resolution,
// and archive extraction (local and remote via SSH pipe).
package main

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

func isInGitRepo() bool {
	cmd := exec.Command("git", "rev-parse", "--is-inside-work-tree")
	cmd.Stderr = nil
	cmd.Stdout = nil
	return cmd.Run() == nil
}

func isTreeClean() bool {
	// Single porcelain call detects staged, unstaged, and untracked changes
	out, err := exec.Command("git", "status", "--porcelain").Output()
	if err != nil {
		return false
	}
	return len(strings.TrimSpace(string(out))) == 0
}

func resolveHEAD() (string, error) {
	return resolveRef("HEAD")
}

// resolveRef resolves a git ref (SHA, branch, HEAD, etc.) to a full SHA.
func resolveRef(ref string) (string, error) {
	out, err := exec.Command("git", "rev-parse", ref).Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// gitRepoRoot returns the top-level directory of the git repository.
func gitRepoRoot() (string, error) {
	out, err := exec.Command("git", "rev-parse", "--show-toplevel").Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// validWorkdir returns cwd if it still exists, otherwise falls back
// to the git repo root. Prevents "cannot get current path" errors
// when cwd is deleted during long-running async handlers.
func validWorkdir() string {
	cwd, err := os.Getwd()
	if err == nil {
		if _, statErr := os.Stat(cwd); statErr == nil {
			return cwd
		}
	}
	root, err := gitRepoRoot()
	if err == nil {
		return root
	}
	return "."
}

// isCommitPushed checks if a SHA exists on any remote branch.
func isCommitPushed(sha string) bool {
	out, err := exec.Command("git", "branch", "-r", "--contains", sha).Output()
	if err != nil {
		return false
	}
	return len(strings.TrimSpace(string(out))) > 0
}

// extractRepoLocal extracts the repo at the given SHA to a local directory.
func extractRepoLocal(sha, dir string) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	untar := exec.Command("tar", "-C", dir, "-x")
	if err := pipeGitArchive(sha, untar); err != nil {
		return err
	}
	return exec.Command("chmod", "-R", "u+w", dir).Run()
}

// extractRepoRemote extracts the repo at the given SHA to a remote host via SSH.
func extractRepoRemote(sha, host, dir string) error {
	sshCmd := exec.Command("ssh", host,
		fmt.Sprintf("mkdir -p '%s' && tar -C '%s' -x && chmod -R u+w '%s'", dir, dir, dir))
	sshCmd.Stdout = os.Stdout
	sshCmd.Stderr = os.Stderr
	return pipeGitArchive(sha, sshCmd)
}

// pipeGitArchive streams `git archive` tar output into a consumer command
// (either local `tar -x` or remote `ssh host "tar -x"`). This avoids
// writing the archive to disk — the tar stream flows through a pipe.
func pipeGitArchive(sha string, consumer *exec.Cmd) error {
	archive := exec.Command("git", "archive", "--format=tar", sha)
	pipe, err := archive.StdoutPipe()
	if err != nil {
		return err
	}
	consumer.Stdin = pipe
	if err := archive.Start(); err != nil {
		return err
	}
	if err := consumer.Run(); err != nil {
		return fmt.Errorf("consumer: %w", err)
	}
	if err := archive.Wait(); err != nil {
		return fmt.Errorf("git archive: %w", err)
	}
	return nil
}
